### ISSUE TITLE
Support for .NET Core stack for azurerm_app_service

### ISSUE THREAD

**Issue by jsve** [hashicorp/terraform-provider-azurerm#5350]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
The provider supports a flag in `site_config` to set the `dotnet_framework_version`. This works well to set one of the .NET Framework versions that are supported by Azure App Services. App Services, however, also supports .NET Core applications since some time last year.

This feature request is to add a possibility to configure app services that run .NET Core.

Azure currently lists these .NET Core versions for the runtime stack: 1.0, 1.1, 2.0, 2.1 (LTS), 2.2, 3.0 (Current), 3.1 (LTS).

### New or Affected Resource(s)

* azurerm_app_service

### Potential Terraform Configuration

```hcl
resource "azurerm_resource_group" "main" {
  name     = "${var.prefix}-resources"
  location = "${var.location}"
}

resource "azurerm_app_service_plan" "main" {
  name                = "${var.prefix}-asp"
  location            = "${azurerm_resource_group.main.location}"
  resource_group_name = "${azurerm_resource_group.main.name}"

  sku {
    tier = "Basic"
    size = "B1"
  }
}

resource "azurerm_app_service" "main" {
  name                = "${var.prefix}-appservice"
  location            = "${azurerm_resource_group.main.location}"
  resource_group_name = "${azurerm_resource_group.main.name}"
  app_service_plan_id = "${azurerm_app_service_plan.main.id}"

  site_config {
    dotnet_core_version = "v3.1"
  }
}
```

### References

* I have not been able to find any current GitHub issues regarding this

### Disclamer / Note for others wanting to deploy Core applications through terraform
Even if you set the `dotnet_framework_version` to "v.4.0" and this is reflected in the web GUI, you will still be able to deploy your Core application to this App Service and they will run regardless of this property.

**Comment by aagoldingay** [comment]
Nice suggestion, that would make it way easier!

Just incase this helps anyone, I was able to set the net core version using the following:

```hcl
resource "azurerm_app_service" "main" {
  name                = "${var.prefix}-appservice"
  location            = "${azurerm_resource_group.main.location}"
  resource_group_name = "${azurerm_resource_group.main.name}"
  app_service_plan_id = "${azurerm_app_service_plan.main.id}"

  site_config {
    linux_fx_version = "DOTNETCORE|3.1"
  }
```

**Comment by phekmat** [comment]
As I understand it, the net core version has no effect on Windows app service plans (aside from setting a metadata field). You also can't specify the version in the portal once the app service is created, only that it's the net core stack. 

On Linux, you use the setting @aagoldingay used.

**Comment by Imed-Eden** [comment]
Hello Everyone,

First of all thank you for your efforts! Do you have any idea when will this feature be added please?

**Comment by bcline760** [comment]
This would be a good thing to have for the App Service resource.

To clarify, .NET Core up to 3.1 will run in an App Service that is configured with the current dotnet_framework_version = "v4.0"? If so, I'll open an issue to have the documentation updated.

**Comment by phekmat** [comment]
> To clarify, .NET Core up to 3.1 will run in an App Service that is configured with the current dotnet_framework_version = "v4.0"? If so, I'll open an issue to have the documentation updated.

It's slightly more nuanced than that, but generally yes for Windows. When *not* using self-contained deployments, the runtime has to be rolled out on the Azure side first. Ditto on the SDK when using kudu to build on the app service. See also https://github.com/Azure/app-service-announcements-discussions/issues/129

**Comment by Imed-Eden** [comment]
Hello,
Are there any updates with regard to dot net core? is it implemented? in the documentation, it is still not there.

**Comment by vivek99singh** [comment]
> Nice suggestion, that would make it way easier!
> 
> Just incase this helps anyone, I was able to set the net core version using the following:
> 
> ```terraform
> resource "azurerm_app_service" "main" {
>   name                = "${var.prefix}-appservice"
>   location            = "${azurerm_resource_group.main.location}"
>   resource_group_name = "${azurerm_resource_group.main.name}"
>   app_service_plan_id = "${azurerm_app_service_plan.main.id}"
> 
>   site_config {
>     linux_fx_version = "DOTNETCORE|3.1"
>   }
> ```

If your app service plan is for linux then linux_fx_version ="DOTNETCORE|3.1" works but it work for windows web app

**Comment by vivek99singh** [comment]
> Hello,
> Are there any updates with regard to dot net core? is it implemented? in the documentation, it is still not there.

I haven't found any update regarding the same. Finding it difficult to find the same even in Azure CLI documentation as well

**Comment by thisisjagan** [comment]
Please advise if there is a way to use .NET 3.1 Stack for Windows App Service in Azure.

**Comment by StefanSchoof** [comment]
@thisisjagan What do mean by "use". If you want to run dotnet core, see this remark in issue text:
> Even if you set the dotnet_framework_version to "v.4.0" and this is reflected in the web GUI, you will still be able to deploy your Core application to this App Service and they will run regardless of this property.

**Comment by StefanSchoof** [comment]
@thisisjagan What do mean by "use". If you want to run dotnet core, see this remark in issue text:
> Even if you set the dotnet_framework_version to "v.4.0" and this is reflected in the web GUI, you will still be able to deploy your Core application to this App Service and they will run regardless of this property.

**Comment by StefanSchoof** [comment]
@thisisjagan What do mean by "use". If you want to run dotnet core, see this remark in issue text:
> Even if you set the dotnet_framework_version to "v.4.0" and this is reflected in the web GUI, you will still be able to deploy your Core application to this App Service and they will run regardless of this property.

**Comment by Jitendrakandpal11** [comment]
> @thisisjagan What do mean by "use". If you want to run dotnet core, see this remark in issue text:
> 
> > Even if you set the dotnet_framework_version to "v.4.0" and this is reflected in the web GUI, you will still be able to deploy your Core application to this App Service and they will run regardless of this property.

It won't work it gives
HTTP Error 500.31 - ANCM Failed to Find Native Dependencies
Common solutions to this issue:
The specified version of Microsoft.NetCore.App or Microsoft.AspNetCore.App was not found.
Troubleshooting steps:
Check the system event log for error messages
Enable logging the application process' stdout messages
Attach a debugger to the application process and inspect
For more information visit: https://go.microsoft.com/fwlink/?LinkID=2028526 error

**Comment by RavulaChetan** [comment]
Any new update on this?
I used terraform to call ARM template as work around about an year and half back. I can't believe this is still not available for windows app service plan.

**Comment by sunnyruziev** [comment]
For Windows App plan use     ' windows_fx_version = "DOTNETCORE|3.1" '

**Comment by scott1138** [comment]
Is there a reason that the documentation hasn't been updated to reflect that .NET Core runtimes are installed on the Windows plan regardless of the site_settings?  I deployed an app using `windows_fx_version = "DOTNETCORE"` as suggested above and it deployed as .NET Framework v4.0.  I changed it to .Net Core in the portal and terraform sees no change in state.  More clarity around this setting would be appreciated.

**Comment by StefanSchoof** [comment]
I think because the core team has many important tasks and nobody of us created a PR to improve the docs. Usually docs PR are merged quite fast. So best thing, to help others not running into this problem is to create a PR to update the docs.

**Comment by scott1138** [comment]
@StefanSchoof Before I update the docs, I'd need to do some more testing and/or get confirmation from the dev team that what we are seeing is WAI and not just a fluke that the .Net Core runtimes are all there.  If it's a function that just hasn't been added yet to the Go SDK and future behavior might change, I don't want to lead people astray.

**Comment by andrew-benson** [comment]
Here's a workaround, add it after your `resource "azurerm_app_service" ` code

```
resource "azurerm_template_deployment" "webapp-corestack" {
  name                = "AspNetCoreStack"
  resource_group_name = azurerm_resource_group.rg.name
  template_body       = <<DEPLOY
  {
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "siteName": {
            "type": "string",
            "metadata": {
                "description": "The Azure App Service Name"
            }
        }
    },
    "resources": [
        {
            "apiVersion": "2018-02-01",
            "name": "[parameters('siteName')]",
            "type": "Microsoft.Web/sites",
            "location": "[resourceGroup().location]",
            "properties": {
                "name": "[parameters('siteName')]",
                "siteConfig": {
                    "appSettings": [],
                    "metadata": [
                        {
                            "name": "CURRENT_STACK",
                            "value": "dotnetcore"
                        }
                    ]
                }
            }
        }
    ]
}
  DEPLOY
  parameters = {
    "siteName"         = azurerm_app_service.webapp.name
  }
  deployment_mode = "Incremental"
  depends_on      = [azurerm_app_service.webapp]
}
```

Helpful article: https://faultbucket.ca/2020/08/terraform-deploy-azure-app-service-with-dotnet-core-stack/

**Comment by surlypants** [comment]
seems "DOTNETCORE|2.1" is not accepted (at least in 2.42.0), whereas "DOTNETCORE|3.1" is :/

**Comment by alastairtree** [comment]
On Windows, with the latest versions of the azurerm provider and what is currently deployed to azure app service, dotnet core apps for 3.1 and 5.0 will work fine using the following configs below. This does not set the "stack" metadata variables but those don't actually seem to change the VMs running your app.

dotnet core 3.1/5.0 apps work and run fine without setting the stack. 5.0 is in [early access](https://azure.github.io/AppService/2020/11/10/Dot-Net-5-on-App-Service.html) at the moment but works fine for me. I have also observeed that is seems 5.0 is rolling out to all servers anyway, and sometimes at random I provision a v4.0 server that already has .net 5 on it without any extra config.

So in summary:

**core 3.1 runs ok when framework is set to 4**
```
resource "azurerm_app_service" "app_service" {
   ...
   site_config {
    dotnet_framework_version = "v4.0"
  }
}
```

**core 5.0 runs ok when framework is set to 5**
```
resource "azurerm_app_service" "app_service" {
   ...
   site_config {
    dotnet_framework_version = "v5.0"
  }
}
```

**Comment by alastairtree** [comment]
On Windows, with the latest versions of the azurerm provider and what is currently deployed to azure app service, dotnet core apps for 3.1 and 5.0 will work fine using the following configs below. This does not set the "stack" metadata variables but those don't actually seem to change the VMs running your app.

dotnet core 3.1/5.0 apps work and run fine without setting the stack. 5.0 is in [early access](https://azure.github.io/AppService/2020/11/10/Dot-Net-5-on-App-Service.html) at the moment but works fine for me. I have also observeed that is seems 5.0 is rolling out to all servers anyway, and sometimes at random I provision a v4.0 server that already has .net 5 on it without any extra config.

So in summary:

**core 3.1 runs ok when framework is set to 4**
```
resource "azurerm_app_service" "app_service" {
   ...
   site_config {
    dotnet_framework_version = "v4.0"
  }
}
```

**core 5.0 runs ok when framework is set to 5**
```
resource "azurerm_app_service" "app_service" {
   ...
   site_config {
    dotnet_framework_version = "v5.0"
  }
}
```

**Comment by tombuildsstuff** [comment]
Worth noting the new opt-in/Preview resources have support for this:

* Linux: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_web_app
* Windows: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/windows_web_app

More information on the Preview can be found in the 3.0 overview guide: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/3.0-overview

**Comment by katbyte** [comment]
Going to close this out as v3.0 of the provider will support this with the new resources!
